Monitor for FloatingIPs to make the resources reservable

AnishReddyRavula opened the pull request:
make the monitor plugin for FLoating IP plugin - mostly tried to make it similar to host monitor plugin

AnishReddyRavula committed:
Monitor for FloatingIPs to make the resources reservable

make the monitor plugin for FLoating IP plugin - mostly tried to make it similar to host monitor plugin

AnishReddyRavula committed:
include neutron client wrapper

Mark-Powers reviewed (commented):
I think there is some confusion about the different floating IP resources and the monitor interface.

Regarding the monitor:
- `set_reservable()` is automatically called based on the returned `failed/recovered` lists. And those two lists should only contain the fips that changed (e.g. are marked as reservable, but actually look unhealthy) For example, in the [host monitor](https://github.com/ChameleonCloud/blazar/blob/chameleoncloud/xena/blazar/plugins/oshosts/host_plugin.py#L952-L954) it only uses the list of recovered hosts is a sublist of the unreservable hosts.

Regarding fip objects:
- there is the floating ip stored in the blazar db FloatingIp table that you added queries for. It seems like you are using this and the allocations to check if an error occured
- the error the user sees most often with floating ip is "floating ip already allocated in subnet". This is referring to neutron floating ips that should've been cleaned up were not.
- So, I don't think you can fully solve the problem just in blazar. You'll need to query neutron for floating IPs from past reservations, and check if they did not get cleaned up. I think the easiest way to do this is look at the `tags` of a neutron floating ip, when a floating IP lease starts, we tag the floating ip with the reservation id [here](https://github.com/ChameleonCloud/blazar/blob/chameleoncloud/xena/blazar/utils/openstack/neutron.py#L91-L93). If that reservation id is from a reservation that is over, it can safely be deleted.

One way to test this, would be to create a lease for a floating ip, and then while the lease is active inspect it with `openstack floating ip show IP_ADDRESS`. Then, the lease will end, and this will be deleted. If you manually ran `openstack floating ip create` with the same details from `floating ip show`, what we want is for this monitor to delete that ip after it executes.

AnishReddyRavula committed:
Clean FIPs if the reservation tagged in neutron is deleted or in error state

process FIPs in parallel through threads

Todo: Add test cases

AnishReddyRavula committed:
Catch exception when the tag does not have blazar reservation

AnishReddyRavula committed:
Unit tests for the floatingip cleanup

AnishReddyRavula committed:
add monitor plugin opts

AnishReddyRavula committed:
change the Db filter to get all the reservations of the list of floating ips

AnishReddyRavula committed:
change set_reservable in fip plugin to update the reservable flag

add floatingip_update db utils method

AnishReddyRavula committed:
Make changes to appropriate cases to return the polled FIPs to recovered or failed

AnishReddyRavula committed:
Get the most recent FIP reservation to skip it when cleaning up

AnishReddyRavula committed:
unittests fix

AnishReddyRavula committed:
Some log messages to understand what is going on

AnishReddyRavula committed:
optimize query to use inner joins and better logging

AnishReddyRavula committed:
use list_floatingips instead of show_floatingip as show might not be working

AnishReddyRavula committed:
Use UTC time and better logging statements with test cases

Use UTC time to filter reservations
Raise FIP not found in neutron wrapper show_floatinip method.
Use appropriate logging debug/warning statements
recover FIP if FIP does not have tags in neutron
New test cases for floatingip plugin monitor

AnishReddyRavula committed:
remove unwanted and condition

AnishReddyRavula committed:
Add dry run flag for so floating IPs are not deleted in neutron

AnishReddyRavula committed:
Use innter join for getting floating ip reservations by FIP id

AnishReddyRavula committed:
COnsider the case when the reservation tag from neutron is not found in blazar DB

Mark-Powers commented on blazar/plugins/floatingips/floatingip_plugin.py:
```diff
@@ -447,3 +455,109 @@ def query_fip_allocations(self, fips, detail=None, lease_id=None,
+                return
+            fip_pool = neutron.FloatingIPPool(fip['floating_network_id'])
+            # check if the FIP is in neutron subnet allocation pools
+            try:
```
This try block I am not sure about. It seems reasonable, but I don't fully understand all of the expected configuration. @msherman64 could you take a look?

msherman64 replied:
this try, except block looks good to me.
that exception indicates that the relevant ipv4 address is within a dhcp allocation pool on the public subnet.

Mark-Powers commented on blazar/plugins/floatingips/floatingip_plugin.py:
```diff
@@ -447,3 +455,109 @@ def query_fip_allocations(self, fips, detail=None, lease_id=None,
+            try:
+                fip_info_from_neutron = fip_pool.show_floatingip(fip["floating_ip_address"])
+            except Exception as e:
+                # If the FIP is not registered in neutron that means there is no reservation for this IP
```
I think this should catch NeutronNotFound or whatever the specific exception is, and the comment should clarify something like:
"If the FIP is not created in neutron, then it is in the correct 'pending' state"
as it's not really about a "no reservation for this IP" but rather the IP is not already in use.

Mark-Powers commented on blazar/plugins/floatingips/floatingip_plugin.py:
```diff
@@ -447,3 +455,109 @@ def query_fip_allocations(self, fips, detail=None, lease_id=None,
+            except Exception as e:
+                # If the FIP is not registered in neutron that means there is no reservation for this IP
+                LOG.debug("Error getting Floating IP from neutron - {e}, recovering...")
+                recovered.append(fip)
```
This shouldn't be added to recovered, as there should be nothing wrong with this case.

Mark-Powers reviewed (changes requested):
This looks great! There are a few minor-ish concerns I have, but I think overall it would run fine.

I am not sure about the use of failed/recovered lists. My understanding based on the host monitor is these lists are meant to set an FIP as reservable/not reservable. That should be done only if neutron is misconfigured, or deleting the ip doesn't work. I think it is fine to add an already reservable ip to the "reservable" list, but that isn't done in the host monitor, and will probably generate a confusing log message like "123.123.123.123 recovered" for IPs that never had problems.

AnishReddyRavula committed:
Use explicit Exception catch when FIP is not found in neutron

Do not add to recoved list when the FIP is already in a good state.

append fip to failed if the FIP cannot be deleted from neutron

AnishReddyRavula committed:
Rename the config opt for polling monitor dry run
